### PR TITLE
Support apis for task dispatcher

### PR DESCRIPTION
**Pull request by youliangtan**
This extends the apis for `rmf_task` and `rmf_battery`, which pertain to the task dispatcher in https://github.com/osrf/rmf_core/pull/217

Example and test scripts:
```bash
ros2 run rmf_fleet_adapter_python test_delivery
ros2 run rmf_fleet_adapter_python test_loop
ros2 run rmf_fleet_adapter_python dispatcher_node
```

### CODE REVIEW

**Comment by youliangtan** [comment]
Update according to https://github.com/osrf/rmf_core/pull/228

File: rmf_fleet_adapter_python/examples/dispatcher_node.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Copyright 2020 Open Source Robotics Foundation, Inc.
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/examples/dispatcher_node.py]:
```suggestion
# Copyright 2021 Open Source Robotics Foundation, Inc.
```

File: rmf_fleet_adapter_python/examples/dispatcher_node.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    print(f" active list >>  {mod.get_active_task_ids()}")
+
+    time.sleep(3)
+    print("---- Assertion ----")
+    state1 = mod.get_task_state(id1)
+    state2 = mod.get_task_state(id2)
+    state3 = mod.get_task_state("null_id")
+    print(f" {id1}: {state1}   \n {id2}: {state2} \n null_id: {state3}")
+
+    assert state1 == Nodes.TaskState.Failed, "state should fail cuz no bid"
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/examples/dispatcher_node.py]:
```suggestion
    assert state1 == Nodes.TaskState.Failed, "state should fail due to absence of a bid"
```

File: rmf_fleet_adapter_python/examples/traffic_light.py [hunk 3]
```diff
@@ -157,6 +161,16 @@ def control_robot(mod, path):
...
 
+    print(f"Sequence of visited waypoints: \n{visited_waypoints}")
+    assert visited_waypoints == [
+        ("dummybot2", 3), 
+        ("dummybot2", 4), 
+        ("dummybot1", 2), 
+        ("dummybot2", 6), 
+        ("dummybot1", 3), 
+        ("dummybot1", 4), 
+        ("dummybot1", 5)], "Sequence seems to be incorrect, pls check"
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/examples/traffic_light.py]:
```suggestion
        ("dummybot1", 5)], "Sequence of visited waypoints is incorrect."
```

File: rmf_fleet_adapter_python/scripts/test_delivery.py [hunk 4]
```diff
@@ -113,7 +115,28 @@ def main():
...
     adapter = adpt.MockAdapter("TestDeliveryAdapter")
     fleet = adapter.add_fleet(fleet_name, robot_traits, test_graph)
-    fleet.accept_delivery_requests(lambda x: True)
+
+    # Set up task request callback function
+    def task_request_cb(task):
+        assert task.task_id == test_name
+        from rmf_task_msgs.msg import TaskType
+        assert task.description.task_type == TaskType.TYPE_DELIVERY
+        print("\nDEBUG!!! ", task_desc.start_time_sec)
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/scripts/test_delivery.py]: Delete if print statement no longer required.

File: rmf_fleet_adapter_python/scripts/test_delivery.py [hunk 5]
```diff
@@ -113,7 +115,28 @@ def main():
     # Manages delivery or loop requests
     adapter = adpt.MockAdapter("TestDeliveryAdapter")
     fleet = adapter.add_fleet(fleet_name, robot_traits, test_graph)
-    fleet.accept_delivery_requests(lambda x: True)
+
+    # Set up task request callback function
+    def task_request_cb(task):
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/scripts/test_delivery.py]: Rename argument `task` to `task_profile` for better readability

Comment (Yadunund) [inline rmf_fleet_adapter_python/scripts/test_delivery.py]: Is there a reason to have assertions inside `task_request_cb`? This should simply be a function that returns `True` if the `task_type == TaskType.TYPE_DELIVERY`. Else return `False`.
Same feedback for function in `test_loop.py`

File: rmf_fleet_adapter_python/src/adapter.cpp [hunk 6]
```diff
@@ -116,8 +133,43 @@ PYBIND11_MODULE(rmf_adapter, m) {
...
+        .def("accept_task_requests",
+             &agv::FleetUpdateHandle::accept_task_requests,
+             py::arg("check"),
+             "Provide a callback function which will accept rmf tasks requests")
+        .def_property("default_maximum_delay",
+             py::overload_cast<>(
+                 &agv::FleetUpdateHandle::default_maximum_delay, py::const_),
+             [&](agv::FleetUpdateHandle& self){
+                 return self.default_maximum_delay();
+             });
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/src/adapter.cpp]: We should also add `fleet_state_publish_period()`
https://github.com/osrf/rmf_core/blob/8d2bea5f5bd222e0ead74540258988e4c30fc074/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/FleetUpdateHandle.hpp#L162-L166

File: rmf_fleet_adapter_python/src/battery/battery.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+      .def_property_readonly("get_capacity",
+        &agv::BatterySystem::capacity)
+      .def_property_readonly("get_charging_current",
+        &agv::BatterySystem::charging_current);
+
+  // motion sink
+  py::class_<agv::SimpleMotionPowerSink>(m_battery, "SimpleMotionPowerSink")
+      // NOTE: MechanicalSystem is being abstracted
+      // .def(py::init<agv::BatterySystem&, agv::MechanicalSystem&>())
+      .def(py::init(
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/src/battery/battery.cpp]: This does not seem to reflect the function signature in `rmf_battery::agv::SimpleMotionPowerSink`. Shouldn't we also be writing bindings for `rmf_battery::agv::MechanicalSystem` and `rmf_battery::agv::PowerSystem`?

File: rmf_fleet_adapter_python/src/battery/battery.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            mass, inertia, friction_coefficient);
+          return std::unique_ptr<agv::SimpleMotionPowerSink>(
+              new agv::SimpleMotionPowerSink(
+                battery_system, *mechanical_system));
+        }),
+        py::arg("battery_system"),
+        py::arg("mass"),
+        py::arg("moment_of_inertia"),
+        py::arg("friction_coefficient"),
+         "Note: MechanicalSystem is being abstracted")
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/src/battery/battery.cpp]: I think abstracting will cause more confusion in this case. We should ensure that the class and its members closely follow the C++ ones

File: rmf_fleet_adapter_python/src/battery/battery.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+          return self.mechanical_system().moment_of_inertia();
+        })
+      .def_property_readonly("get_friction_coefficient",
+        [&](agv::SimpleMotionPowerSink &self){
+          return self.mechanical_system().friction_coefficient();
+        });
+
+  // ambient sink and tool sink
+  py::class_<agv::SimpleDevicePowerSink>(m_battery, "SimpleDevicePowerSink")
+      // NOTE: PowerSystem is being abstracted
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/src/battery/battery.cpp]: Same feedback as above

File: rmf_fleet_adapter_python/src/types/types.cpp [hunk 10]
```diff
@@ -1,3 +1,20 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/src/types/types.cpp]:
```suggestion
 * Copyright (C) 2021 Open Source Robotics Foundation
```

File: rmf_fleet_adapter_python/src/types/types.cpp [hunk 11]
```diff
@@ -16,134 +33,178 @@
...
+            [&](Loop &self) { return self.start_name; },
+            [&](Loop &self,
+                std::string start_name) { self.start_name = start_name; })
+        .def_property(
+            "finish_name",
+            [&](Loop &self) { return self.finish_name; },
+            [&](Loop &self,
+                std::string finish_name) { self.finish_name = finish_name; });
+
+    // TODO: Support More task_types
```
Comment (Yadunund) [inline rmf_fleet_adapter_python/src/types/types.cpp]: We should also include `Clean` task here since the rest of the fleet adapter supports it.

**Review by Yadunund — CHANGES_REQUESTED** [review]
Thanks for updating the bindings. Have left some feedback in the comments below. One other point to note is that the graph class needs to be updated to include [these](https://github.com/osrf/rmf_core/blob/8d2bea5f5bd222e0ead74540258988e4c30fc074/rmf_traffic/include/rmf_traffic/agv/Graph.hpp#L82-L88) functions that were added.

File: rmf_fleet_adapter_python/scripts/test_delivery.py (hunk 5 above)
Comment (youliangtan) [reply rmf_fleet_adapter_python/scripts/test_delivery.py]: This is assertion is to check if the callback function's argument is providing the same task as requested via `dispatch_task()`. This mainly serves as an example to the user. if needed, I can add some comments to elaborate on it.

File: rmf_fleet_adapter_python/src/battery/battery.cpp (hunk 7 above)
Comment (youliangtan) [reply rmf_fleet_adapter_python/src/battery/battery.cpp]: Here, the `MechanicalSystem` and `PowerSystem` are both being abstracted within `SimpleMotionPowerSink` and `SimpleDevicePowerSink` respectively. This reduces the number of class, simplied the setup of battery for the user. e.g: 
```
battery_sys = battery.BatterySystem.make(24.0, 40.0, 8.8)
motion_sink = battery.SimpleMotionPowerSink(battery_sys, 70.0, 40.0, 0.22)
ambient_sink = battery.SimpleDevicePowerSink(battery_sys, 20.0)
tool_sink = battery.SimpleDevicePowerSink(battery_sys, 10.0)
```

Comment (Yadunund) [reply rmf_fleet_adapter_python/src/battery/battery.cpp]: I understand the intention here. But the purpose of the bindings would be to mimic the original C++ API as closely as possible. These classes may simply be containers for physical properties, but they could be expanded in the future to have other functions. Another point is that if the documentation here is lacking, the user may refer to the C++ documentation for these classes. So we do not want to have any discrepancy between the two.

Comment (youliangtan) [reply rmf_fleet_adapter_python/src/battery/battery.cpp]: Updated: https://github.com/osrf/rmf_fleet_adapter_python/pull/10/commits/e6db6ca49713634c0c93e5fa12987faefcb31841

**Comment by Yadunund** [comment]
@tanyouliang95, can we address this point before we can merge.
> One other point to note is that the graph class needs to be updated to include [these](https://github.com/osrf/rmf_core/blob/8d2bea5f5bd222e0ead74540258988e4c30fc074/rmf_traffic/include/rmf_traffic/agv/Graph.hpp#L82-L88) functions that were added.

**Review by Yadunund — APPROVED** [review]
Thanks for addressing all the issues. LGTM!

**Pull request merged**